core/services/relay/delegate.go

```diff
@@ -24,185 +19,114 @@ var (
-type OCR2ProviderArgs struct {
-	ID              int32
-	ContractID      string
-	TransmitterID   null.String
-	Relay           types.Network
-	RelayConfig     job.JSONConfig
-	IsBootstrapPeer bool
-	Plugin          job.OCR2PluginType
-}
-
-// NewOCR2Provider creates a new OCR2 provider instance.
-func (d delegate) NewOCR2Provider(externalJobID uuid.UUID, s interface{}) (types.OCR2ProviderCtx, error) {
-	// We expect trusted input
-	spec := s.(*OCR2ProviderArgs)
-	choice := spec.Relay
-	switch choice {
+func (d Delegate) NewConfigWatcher(relay types.Network, args types.ConfigWatcherArgs) (types.ConfigWatcher, error) {
+	switch relay {
 	case types.EVM:
 		r, exists := d.relayers[types.EVM]
 		if !exists {
 			return nil, errors.New("no EVM relay found; is EVM enabled?")
 		}
-
-		var config evm.RelayConfig
-		err := json.Unmarshal(spec.RelayConfig.Bytes(), &config)
-		if err != nil {
-			return nil, err
-		}
-
-		return r.NewOCR2Provider(externalJobID, evm.OCR2Spec{
-			ID:            spec.ID,
-			IsBootstrap:   spec.IsBootstrapPeer,
-			ContractID:    spec.ContractID,
-			TransmitterID: spec.TransmitterID,
-			ChainID:       config.ChainID.ToInt(),
-			Plugin:        spec.Plugin,
-		})
+		return r.NewConfigWatcher(args)
 	case types.Solana:
 		r, exists := d.relayers[types.Solana]
 		if !exists {
 			return nil, errors.New("no Solana relay found; is Solana enabled?")
 		}
-
-		var config solana.RelayConfig
-		err := json.Unmarshal(spec.RelayConfig.Bytes(), &config)
-		if err != nil {
-			return nil, errors.Wrap(err, "error on 'spec.RelayConfig' unmarshal")
-		}
-
-		// use state account as contract ID (unique for each feed, program account is not)
-		stateID, err := solanaGo.PublicKeyFromBase58(spec.ContractID)
-		if err != nil {
-			return nil, errors.Wrap(err, "error on 'solana.PublicKeyFromBase58' for 'spec.ContractID")
-		}
-
-		programID, err := solanaGo.PublicKeyFromBase58(config.OCR2ProgramID)
-		if err != nil {
-			return nil, errors.Wrap(err, "error on 'solana.PublicKeyFromBase58' for 'spec.RelayConfig.OCR2ProgramID")
-		}
-
-		storeProgramID, err := solanaGo.PublicKeyFromBase58(config.StoreProgramID)
-		if err != nil {
-			return nil, errors.Wrap(err, "error on 'solana.PublicKeyFromBase58' for 'spec.RelayConfig.StateID")
+		return r.NewConfigWatcher(args)
+	case types.Terra:
+		r, exists := d.relayers[types.Terra]
+		if !exists {
+			return nil, errors.New("no Terra relay found; is Terra enabled?")
 		}
+		return r.NewConfigWatcher(args)
+	default:
+		return nil, errors.Errorf("unknown relayer network type: %s", relay)
+	}
+}
 
-		transmissionsID, err := solanaGo.PublicKeyFromBase58(config.TransmissionsID)
-		if err != nil {
-			return nil, errors.Wrap(err, "error on 'solana.PublicKeyFromBase58' for 'spec.RelayConfig.TransmissionsID")
+func (d Delegate) NewMedianProvider(relay types.Network, args types.PluginArgs) (types.MedianProvider, error) {
+	switch relay {
```

Comment:
Could we extract a "getRelay" method for the duplicated switch logic and simply call `NewMedianProvider` and `NewConfigWatcher` on the result?